### ISSUE TITLE
How to select parent that contains certain children from his ManyToMany relation? 

### ISSUE THREAD

**Issue by Reszkettle** [tortoise/tortoise-orm#841]
Given this piece of code:

`class Recipe(Model):
    description = fields.CharField(max_length=1024)
    ingredients = fields.ManyToManyField(model_name="models.Ingredient", on_delete=fields.SET_NULL)

class Ingredient(Model):
    name = fields.CharField(max_length=128)
`

How to query all recipes that contain **BOTH** Ingredient.name="tomato" and Ingredient.name="onion" ?

**Comment by Reszkettle** [comment]
I've tried to make it this way, but still without luck:
    `print(await Recipe.filter(Q(ingredients__name="potato") & Q(id__in=Subquery(Recipe.filter(ingredients__name="tomato").values("id")))))
`

**Comment by dstlny** [comment]
> I've tried to make it this way, but still without luck:
> 
> ```
> qs_potato = Q(ingredients__name="potato")
> qs_tomato = Q(id__in=Subquery(Recipe.filter(ingredients__name="tomato").values("id")))
> await Recipe.filter(qs_potato & qs_tomato)
> ```

```python
qs_potato = Q(ingredients__name="potato")
qs_tomato = Q(ingredients__name="tomato")
await Recipe.filter(as_potato | qs_tomato)

**Comment by Reszkettle** [comment]
I want recipes that contain BOTH ingredients, not one of them

**Comment by Reszkettle** [comment]
> > I've tried to make it this way, but still without luck:
> > 
> > ```
> > qs_potato = Q(ingredients__name="potato")
> > qs_tomato = Q(id__in=Subquery(Recipe.filter(ingredients__name="tomato").values("id")))
> > await Recipe.filter(qs_potato & qs_tomato)
> > ```
>  As simple as: 
> ```python
> qs_potato = Q(ingredients__name="potato")
> qs_tomato = Q(ingredients__name="tomato")
> await Recipe.filter(as_potato | qs_tomato)
> ## or 
> await Recipe.filter(Q(qs_potato, qs_tomato, join_type=Q.OR))
> ```
> 

I want recipes that contain BOTH ingredients, not one of them

**Comment by dstlny** [comment]
> > > I've tried to make it this way, but still without luck:
> > > ```
> > > qs_potato = Q(ingredients__name="potato")
> > > qs_tomato = Q(id__in=Subquery(Recipe.filter(ingredients__name="tomato").values("id")))
> > > await Recipe.filter(qs_potato & qs_tomato)
> > > ```
> > 
> > 
> > As simple as:
> > ```python
> > qs_potato = Q(ingredients__name="potato")
> > qs_tomato = Q(ingredients__name="tomato")
> > await Recipe.filter(as_potato | qs_tomato)
> > ## or 
> > await Recipe.filter(Q(qs_potato, qs_tomato, join_type=Q.OR))
> > ```
> 
> I want recipes that contain BOTH ingredients, not one of them

Woops, misread initially. In that case, you'd probably want to do something like

```python
 _potato = Q(name="potato")
_tomato = Q(name="tomato")
await Recipe.filter(ingredients__in=await Ingredient.filter(Q(_potato, _tomato, join_type=Q.OR)))
```

That way, you're only getting Recipe's which have both those Ingredients.

Could probably turn the ingredients for the Recipe into a subquery to optimise it, but you get the drift :).

**Comment by Reszkettle** [comment]
> > > > I've tried to make it this way, but still without luck:
> > > > ```
> > > > qs_potato = Q(ingredients__name="potato")
> > > > qs_tomato = Q(id__in=Subquery(Recipe.filter(ingredients__name="tomato").values("id")))
> > > > await Recipe.filter(qs_potato & qs_tomato)
> > > > ```
> > > 
> > > 
> > > As simple as:
> > > ```python
> > > qs_potato = Q(ingredients__name="potato")
> > > qs_tomato = Q(ingredients__name="tomato")
> > > await Recipe.filter(as_potato | qs_tomato)
> > > ## or 
> > > await Recipe.filter(Q(qs_potato, qs_tomato, join_type=Q.OR))
> > > ```
> > 
> > I want recipes that contain BOTH ingredients, not one of them
> 
> Woops, misread initially. In that case, you'd probably want to do something like
> 
> ```python
>  _potato = Q(name="potato")
> _tomato = Q(name="tomato")
> await Recipe.filter(ingredients__in=await Ingredient.filter(Q(_potato, _tomato, join_type=Q.OR)))
> ```
> 
> That way, you're only getting Recipe's which have both those Ingredients.
> 
> Could probably turn the ingredients for the Recipe into a subquery to optimise it into a single query w/ a sub-query instead of two distinct queries, but you get the drift :).

Thanks in advance. Gonna check it once I reach my PC :D!

**Comment by Reszkettle** [comment]
By the way... maybe I'm not good in understanding how this ORM work, but just by reading what you have typed, won't that actually query again Recipes that have Tomato or Potato, will it? Cause there is this "in" operator, which as far as I know, works as this "in" operator from raw SQL

**Comment by dstlny** [comment]
> By the way... maybe I'm not good in understanding how this ORM work, but just by reading what you have typed, won't that actually query again Recipes that have Tomato or Potato, will it? Cause there is this "in" operator, which as far as I know, works as this "in" operator from raw SQL

Nah, it'll just be two queries. One query to get the Ingredients with those names, and then another to get Recipes which have those two Ingredients via the through table linking those two tables (for the M2M to work) :)

**Comment by Reszkettle** [comment]
Yeah it works pe

> 
> 
> > By the way... maybe I'm not good in understanding how this ORM work, but just by reading what you have typed, won't that actually query again Recipes that have Tomato or Potato, will it? Cause there is this "in" operator, which as far as I know, works as this "in" operator from raw SQL
> 
> Nah, it'll just be two queries. One query to get the Ingredients with those names, and then another to get Recipes which have those two Ingredients via the through table linking those two tables (for the M2M to work) :)

Yeah it works like a charm, perfect, thank you!
However could you tell me what is the difference between:
`await Ingredient.filter(Q(_potato, _tomato, join_type=Q.OR))` and 
`await Ingredient.filter(_potato | _tomato)`

**Comment by Reszkettle** [comment]
> 
> 
> > By the way... maybe I'm not good in understanding how this ORM work, but just by reading what you have typed, won't that actually query again Recipes that have Tomato or Potato, will it? Cause there is this "in" operator, which as far as I know, works as this "in" operator from raw SQL
> 
> Nah, it'll just be two queries. One query to get the Ingredients with those names, and then another to get Recipes which have those two Ingredients via the through table linking those two tables (for the M2M to work) :)

It didn't work it gives me as I expected both the Recipe with only one ingredient, and Recipe with both ingredients :/

**Comment by Reszkettle** [comment]
` # Creating ingredients
    tomato = await Ingredient.create(name="tomato")
    onion = await Ingredient.create(name="onion")

    # Creating 1st recipe
    first_recipe = await Recipe.create(description="This is the 1st recipe")
    await first_recipe.ingredients.add(tomato)

    # Creating 2nd recipe
    second_recipe = await Recipe.create(description="This is the 2nd recipe")
    await second_recipe.ingredients.add(tomato)
    await second_recipe.ingredients.add(onion)

    _tomato = Q(name="tomato")
    _onion = Q(name="onion")
    print(await Recipe.filter(ingredients__in=await Ingredient.filter(Q(_tomato, _onion, join_type=Q.OR))))`

**Comment by Reszkettle** [comment]
This query worked:
`await Recipe.filter(pk__in=Subquery(Recipe.filter(ingredients__name="onion").values("id")), ingredients__name="tomato")`
But I think that this ORM should help me with simplifying it a bit :D

**Comment by spacemanspiff2007** [comment]
@Reszkettle 
I am using the following [snippet](https://github.com/tortoise/tortoise-orm/issues/756#issuecomment-881448673) and it works very well.

**Comment by Reszkettle** [comment]
> @Reszkettle 
> I am using the following [snippet](https://github.com/tortoise/tortoise-orm/issues/756#issuecomment-881448673) and it works very well.
> 
> Usage would be
> ```python
> query = CriterionQuerySet(MyModel)
> query.has_relations_with('ingredients', [Ingredient['onion'], Ingredient['tomato'])
> ```

how about performance side?

**Comment by spacemanspiff2007** [comment]
What do you mean? It's a normal query so not exceptionally slow/fast however I did not bench it.
